lib/JSON/Validator/Util.pm

```diff
@@ -24,7 +24,7 @@ sub E { JSON::Validator::Error->new(@_) }
 my $serializer = SEREAL_SUPPORT ? \&_sereal_encode : \&_yaml_dump;
 
 sub data_checksum {
-  return Mojo::Util::md5_sum(ref $_[0] ? $serializer->($_[0]) : defined $_[0] ? qq('$_[0]') : 'undef');
+  return Mojo::Util::md5_sum(ref $_[0] ? $serializer->($_[0]) : defined $_[0] ? do { utf8::encode(my $x = shift); qq('$x') } : 'undef');
```

Comment:
Could you use encode from Mojo::Util instead? I also wonder what would happen if we run everything through $serializer...

Reply:
I could, but chose specifically to use utf8::encode specifically. It shouldn't be possible for invalid codepoints (e.g., `\x{d800}`) to appear, but if they somehow did Mojo::Util (or Encode directly) would insert replacement characters. Different inputs (with invalid codepoints in the same place) would hash the same and compare the same. With utf8::encode, the perl representation is directly put into codepoints in the resulting string even if they are invalid. This way, different perl strings will always hash different.

If using Mojo::Util::encode

Reply:
If that's the case then you have to add a(nother) test for that, or else I can pretty much guarantee you that I will change this later on and introduce a regression.

Reply:
gotcha, test added

Reply:
FWIW, I ran the benchmark test with n=1000 running this vs. just using serializer for everything. Using serializer was something like 35% slower in validating (just judging by repored time from the test).

Reply:
Thanks for checking! I was expecting something like that, but I wasn't sure.